app/services/activitypub/fetch_replies_service.rb

```diff
@@ -24,18 +28,29 @@ def collection_items(collection_or_uri)
     collection = fetch_collection(collection['first']) if collection['first'].present?
     return unless collection.is_a?(Hash)
 
-    case collection['type']
-    when 'Collection', 'CollectionPage'
-      as_array(collection['items'])
-    when 'OrderedCollection', 'OrderedCollectionPage'
-      as_array(collection['orderedItems'])
+    all_items = []
+    while collection.is_a?(Hash)
+      items = case collection['type']
+              when 'Collection', 'CollectionPage'
+                collection['items']
+              when 'OrderedCollection', 'OrderedCollectionPage'
+                collection['orderedItems']
+              end
+
+      all_items.concat(as_array(items))
+
+      break if all_items.size > MAX_REPLIES
+
+      collection = collection['next'].present? ? fetch_collection(collection['next']) : nil
```

Comment:
I think an infinite loop is possible if an attacker builds a collection of infinitely many empty pages.

Reply:
Ah good point. Should I add a global loop limit here as a classvar? Or should it be a method argument with a default?

Or should we quit if we get an empty collection?

Reply:
[ActivityStreams](https://www.w3.org/TR/activitystreams-core/#collections) doesn't have a recommendation for malicious paging, just says to "page until you get to the end" 

[ActivityPub](https://www.w3.org/TR/activitypub/#security-c2s-response-dos) has a comment re: clients expecting to handle DoS from broken/malicious servers.

so doesn't look like we have a "standards-compliant" call to make here, i think quitting on an empty page makes sense? if someone is feeding us an empty page that's not the `last` page, then that is incorrect behavior and we shoudn't expect objects in future pages. 

the edge case would be a malicious server that wants to serve us 1000 pages of single objects, so should we have a separate page limit aside from the max status limit?

Reply:
> i think quitting on an empty page makes sense? if someone is feeding us an empty page that's not the last page, then that is incorrect behavior and we shoudn't expect objects in future pages.

there is the possibility that an empty page can be served if the server in question uses "stable paging" -- in this case, imagine the following scenario:

- the server has an OrderedCollection sorted reverse chronologically
- the server has a page size of 10
- 27 items get added to the collection

the representation of the collection and its pages now looks something like this:

```
{
3: [ 27, 26, 25, 24, 23, 22, 21 ],
2: [ 20, 19, 18, 17, 16, 15, 14, 13, 12, 11],
1: [ 10, 09, 08, 07, 06, 05, 04, 03, 02, 01 ]
}

OrderedCollection.first = 3
OrderedCollection.last = 1
3.next = 2
2.next = 1
1.prev = 2
2.prev = 3
```

now say you Remove item 7:

```
{
3: [ 27, 26, 25, 24, 23, 22, 21 ],
2: [ 20, 19, 18, 17, 16, 15, 14, 13, 12, 11],
1: [ 10, 09, 08, 06, 05, 04, 03, 02, 01 ]
}

OrderedCollection.first = 3
OrderedCollection.last = 1
3.next = 2
2.next = 1
1.prev = 2
2.prev = 3
```

instead of reflowing the pages so that the 1st page contains 10 items, the use of "stable paging" means that the 1st page is now 9 items instead of 10 items. but in exchange for that, pages 2 and 3 are unchanged. you can surely imagine how problematic it would be for caching if every removal or deletion invalidated the cache of every single page after it, in a cascade.

if all items are removed from a "stable page", you have the chance of ending up with a "degenerate page". consider what happens when you remove items 11-20:

```
{
3: [ 27, 26, 25, 24, 23, 22, 21 ],
2: [ ],
1: [ 10, 09, 08, 06, 05, 04, 03, 02, 01 ]
}

OrderedCollection.first = 3
OrderedCollection.last = 1
3.next = 2
2.next = 1
1.prev = 2
2.prev = 3
```

now, page 1 contains 9 items, page 2 contains 0 items, and page 3 contains 7 items.

> the edge case would be a malicious server that wants to serve us 1000 pages of single objects, so should we have a separate page limit aside from the max status limit?

the "out" here is that you need to:

- keep track of visited page ids, so that you never revisit a page that you've already "seen". this prevents loops.
- set a "max recursion limit" for how many pages you visit before you stop. this prevents unbounded recursion.

one other thing to consider is that perhaps this doesn't need to happen automatically; perhaps it could be triggered by a user action, so that the user can request "show me more replies" and the user-agent (the instance, acting as an activitypub client) can go fetch those replies on demand. the parameters of this action would be:

- depth. reddit for example i think uses a parameter `?context` which defaults to 3, so by default it will show 3 replies upward or downward. any more than that, and the "see more replies" link will be shown below as a child.
- width. this is how many direct descendants to show. i don't know what a good limit is, but using reddit as an example again, it seems to be anywhere from 3-100 depending on the depth in the tree. when the limit gets exceeded, you see a different "see more replies" button", which is placed at the same level alongside the replies.

a final thing to consider is that it is probably simpler to introduce the concept of a "conversation" entity (or broaden the existing one from being narrowly scoped to mentions-only statuses) which contains all statuses in the conversation (ideally decided by a conversation's owner), then show this as a flat chronological list, with the `inReplyTo` being metadata shown above each status. but this is something that should happen more generally and broadly outside of this PR.

Reply:
Alright good point, so we can't just quit on empty pages. Maybe we can have some lower threshold for `n` empty pages in a row though. 

> keep track of visited page ids, so that you never revisit a page that you've already "seen". this prevents loops.

that would be easy to do here, and a good idea too

> set a "max recursion limit" for how many pages you visit before you stop. this prevents unbounded recursion.

ya basically this is what i was asking about above. so maybe some combination of the two values of `n` blank pages and `m` max pages.

> one other thing to consider is that perhaps this doesn't need to happen automatically; perhaps it could be triggered by a user action

ha yes [i wish `context` was paginated too](https://github.com/mastodon/mastodon/issues/11029#issuecomment-2423630034). Since this is a backend service, it's find to fetch all the replies that need fetching, the decision is basically how often to do that. if user intent is to look closer at a status, maybe they also want to eg. search among the replies, so having a reasonable approximation of "all" the replies matches that, and also meets the stated needs that people want to see the same thing as other people on different instances, modulo blocks/mutes/filters/etc. 

what gets sent to the client would ideally be paged tho and i think you're right a shallow broad paging mechanism for context trees would be the right call.

> a final thing to consider is that it is probably simpler to introduce the concept of a "conversation" entity

ya agreed seems unrelated to this pr

Reply:
I think having a low recursion limit makes sense. Every request is costly (up to a few seconds of waiting on a remote server), and the more we do in the service, the longer we will keep a worker busy. This can easily become a DoS vector.

As for paginating the context, yes, that would be extremely useful, and it could be useful in this specific case as well, but as you may have noticed this is a massively complex issue that also involves a lot of UX considerations (for instance, Mastodon's display of replies is “flattened” so the pagination cutoffs are far less obvious to represent than on say, reddit)

Reply:
> I think having a low recursion limit makes sense.

Checking on how we are meaning recursion here, just because activitystreams also refers to just iterating through collection pages as recursion: do we mean a limit on paging or do we mean limiting the recursion depth within a reply tree (only get e.g. `depth<n` replies)?

To me a limit on paging makes perfect sense as a complementary limit to the max statuses limit, but a depth in reply tree recursion is in my opinion undesirable: In practice i have been seeing a pretty even distribution between deep and broad trees, and since there is no paging in the context endpoint, the experience for both being "i go to the status, and a few seconds later I refresh, and all the replies are there" is ... great! Since the action is implicit on loading the context endpoint, and we don't have the ux pieces in place for "see more replies" and recursively expanding trees, I would think that some invisible barrier where someone finds out later 'oh i wasn't actually seeing all the replies, i needed to click around more' would be pretty frustrating.

I think the max page limit handles concerns re: runtime/resource usage in either case of deep and broad trees well, and does it more directly - we care about the number of requests/pages, not the depth of trees per se, the depth of trees is just a proxy for that.

> Every request is costly (up to a few seconds of waiting on a remote server), and the more we do in the service, the longer we will keep a worker busy

For the sake of having it in thread (assuming you have read it, but for the benefit of passers-by the haven't), the way the service works is that there is one long-running `FetchAllRepliesWorker`. That is just getting a list of URIs to fetch from the replies collections. It then hands off those URIs in batches to sets of `FetchReplyWorker`s. So the long-running worker should be mostly I/O bound as you say. We have found it necessary to [modify the sidekiq queues](https://wiki.neuromatch.social/Performance#Sidekiq) when we were as small as a few dozen users, so that might be a thing worth exploring making default separately if we are concerned about long running workers clogging up slots. Otherwise, it makes sense to me for this worker to be one of the main things that occupies the `pull` queue, as most of the rest of the things that use it are even more background-oriented tasks than this is - the one that we see running most commonly are the `SynchronizeFeatured*Collection` workers.

I think this makes sense to make configurable rather than setting a hard limit - for small instances, they probably want to have a higher limit because they are the ones most impacted by not seeing replies, don't have a million users requesting threads, and can let a worker linger for longer. for large instances, they probably want to set the page limit lower, etc.

Reply:
> Checking on how we are meaning recursion here, just because activitystreams also refers to just iterating through collection pages as recursion: do we mean a limit on paging or do we mean limiting the recursion depth within a reply tree (only get e.g. `depth<n` replies)?

I meant a pagination limit, not specifically a reply tree depth limit.

> I would think that some invisible barrier where someone finds out later 'oh i wasn't actually seeing all the replies, i needed to click around more' would be pretty frustrating.

This is always a risk, whatever limits we have, unfortunately.

Reply:
ok yes i'll put in the limits tmrw :)

also yes definitely lmk if there is a better way to do scheduling and how to avoid clogging smaller sidekiq queues. i looked into some other async strategies that seemed like they would allow a worker to pause while awaiting from a batch of jobs, but it looks like its sidekiq pro only. if there is another long-running worker that i should mimic also lmk.

Reply:
completed in https://github.com/mastodon/mastodon/pull/32615/commits/e991fb890cebcc0b33450e76d6decbf6023df4c4 i think